src/shogun/mathematics/graph/ops/Add.h

```diff
@@ -0,0 +1,274 @@
+		{
+			m_output_tensors[0]->data() = allocator_dispatch(
+			    m_output_tensors[0]->get_size_from_shape(shape), type);
+		}
+	};
+
+
+	template <typename EngineImplementation>
+	IGNORE_IN_CLASSLIST class AddImpl: public OperationImpl
+	{
+	public:
+		AddImpl(const std::shared_ptr<Node>& node): OperationImpl(node) {}
+
+		virtual ~AddImpl() {}
+
+		template<typename T>
+		void evaluate()
+		{
+			this->evaluate_implementation();
+		}
+	};
+
+
+	IGNORE_IN_CLASSLIST class AddShogun : public AddImpl<AddShogun>
+	{
+	public:
+		AddShogun(const std::shared_ptr<Node>& node) : AddImpl(node)
+		{
+		}
+
+		void build()
+		{
+		}
+
+		void evaluate()
+		{
+			auto add_node = std::static_pointer_cast<Add>(m_abstract_node);
+
+			const auto& node1 = m_abstract_node->get_input_nodes()[0];
+			const auto& node2 = m_abstract_node->get_input_nodes()[1];
+
+			add_node->allocate_tensor(
+			    rutime_shape_check(node1, node2),
+			    node1->get_tensors()[0]->get_type());
+
+			auto input_tensor1 = node1->get_tensors()[0];
+			auto input_tensor2 = node2->get_tensors()[0];
+			auto output_tensor = add_node->get_tensors()[0];
+
+
+			// the actual call
+			kernel(
+			    input_tensor1->data(),
+			    input_tensor2->data(),
+			    output_tensor->data(), output_tensor->size(),
+			    output_tensor->get_type());
+		}
+
+	private:
+
+		const Shape& rutime_shape_check(const std::shared_ptr<Node>& node1, 
+			const std::shared_ptr<Node>& node2)
+		{
+			// we don't need to check how many tensors there are
+			// the compile time check already did that
+			// we also know that the number of dimensions match
+			const auto& node1_tensor = node1->get_tensors()[0];
+			const auto& node2_tensor = node2->get_tensors()[0];
+
+			for (const auto& [idx, shape1, shape2] : enumerate(
+			         node1_tensor->get_shape(),
+			         node2_tensor->get_shape()))
+			{
+				if (shape1 != shape2)
+				{
+					error("Runtime shape mismatch in dimension {}. Got {} and {}.",
+						idx, shape1, shape2);
+				}
+			}
+
+			return node1_tensor->get_shape();
+		}
+
+		void kernel(
+		    void* input1, void* input2, void* output, size_t size,
+		    element_type type)
+		{
+			switch (type)
+			{
+			case element_type::FLOAT32:
+				kernel_implementation<
+				    get_type_from_enum<element_type::FLOAT32>::type>(
+				    input1, input2, output, size);
+				break;
+			case element_type::FLOAT64:
+				kernel_implementation<
+				    get_type_from_enum<element_type::FLOAT64>::type>(
+				    input1, input2, output, size);
+				break;
+			}
+		}
+
+		template <typename T>
+		void kernel_implementation(
+		    void* input1, void* input2, void* output, size_t size)
+		{
+			// if we have SYCL or MSVC we could add parallel execution
+			// or just use Eigen here
+			std::transform(
+			    static_cast<const T*>(input1),
+			    static_cast<const T*>(input1) + size,
+			    static_cast<const T*>(input2), static_cast<T*>(output),
+			    std::plus<T>());
+		}
+	};
+
+#ifdef USE_NGRAPH
```

Comment:
i guess this could be moved to a separate file and directory that is only related to ngraph and then have a shared lib defined like: shogun-ngraph.so

Reply:
indeed, I started doing that `graph/ops/shogun/OperatorName.h` and so on